src/main/scala/com/softwaremill/realworld/users/UsersService.scala

```diff
@@ -58,19 +58,11 @@ class UsersService(authService: AuthService, usersRepository: UsersRepository):
     val emailCleanOpt = updateData.email.map(email => email.toLowerCase.trim())
     val usernameCleanOpt = updateData.username.map(username => username.trim())
 
-    def checkUserDoesNotExist(emailOpt: Option[String], usernameOpt: Option[String]): IO[Exception, Unit] =
-      (emailOpt, usernameOpt) match
-        case (Some(email), None)    => checkUserDoesNotExistByEmail(email)
-        case (None, Some(username)) => checkUserDoesNotExistByUsername(username)
-        case (Some(email), Some(username)) =>
-          checkUserDoesNotExistByEmail(email)
-          checkUserDoesNotExistByUsername(username)
-        case (None, None) => ZIO.unit
+    def checkUserDoesNotExist(emailOpt: Option[String], usernameOpt: Option[String]): Task[Unit] =
+      ZIO.foreach(emailOpt)(checkUserDoesNotExistByEmail).flatMap(_ => ZIO.foreach(usernameOpt)(checkUserDoesNotExistByUsername).unit)
 
-    def validateEmailOpt(emailCleanOpt: Option[String]): IO[Exception, Unit] =
-      emailCleanOpt match
-        case Some(email) => validateEmail(email)
-        case None        => ZIO.unit
+    def validateEmailOpt(emailCleanOpt: Option[String]): Task[Unit] =
+      ZIO.foreach(emailCleanOpt)(validateEmail).unit
 
     for {
       _ <- validateEmailOpt(emailCleanOpt)
```

Comment:
**nitpick:** Actually, since validation checks have become so simple, I would just put them in the for-comp (and you can skip .unit, becase you're ignoring the result with `_ <-`):

```scala
for {
  _ <- ZIO.foreach(emailCleanOpt)(validateEmail)
  _ <- ZIO.foreach(emailOpt)(checkUserDoesNotExistByEmail)
  _ <- ZIO.foreach(usernameOpt)(checkUserDoesNotExistByUsername)
oldUser <- usersRepository
    .findUserWithPasswordById(userId)
    .someOrFail(NotFound(UserWithIdNotFoundMessage(userId)))
// ...
```